gcp-devel/main.go

```diff
@@ -0,0 +1,333 @@
+	NodeLocations []string
+	NodeConfig    ConfigClustersNodeConfig
+	Names         []string
+}
+
+type Config struct {
+	Clusters ConfigClusters
+	Location string
+}
+
+// GenerateSSHKeys ...
+func GenerateSSHKeys() (string, string, error) {
+	// IMPORTANT: For GCP it has to be 3072 bit key, when tried to use 2048 or
+	//            4096 it was not adding it to the authorized_keys file on
+	//            cluster nodes, even if the key was present in a gcp console
+	//            (in node's metadata)
+	priv, err := rsa.GenerateKey(rand.Reader, 3072)
+	if err != nil {
+		return "", "", err
+	}
+
+	b := pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(priv),
+	}
+
+	privBuffer := bytes.NewBuffer([]byte{})
+	if err := pem.Encode(privBuffer, &b); err != nil {
+		return "", "", err
+	}
+
+	public, err := ssh.NewPublicKey(&priv.PublicKey)
+	if err != nil {
+		return "", "", err
+	}
+	// f, err := os.OpenFile(privKeyPath, os.O_CREATE|os.O_TRUNC|os.O_RDWR, 0400)
+	// if err != nil {
+	// 	return err
+	// }
+	//
+	// defer f.Close()
+
+	return privBuffer.String(), base64.StdEncoding.EncodeToString(public.Marshal()), nil
+}
+
+var (
+	PrivateKeySecretName = "kuma-main-ssh-private-key"
+	PublicKeySecretName  = "kuma-main-ssh-public-key"
+)
+
+func genName(value ...string) string {
+	if DefaultNamePrefix == "" {
+		return strings.Join(value, "-")
+	}
+
+	return strings.Join(append([]string{DefaultNamePrefix}, value...), "-")
+}
+
+func main() {
+	u, err := user.Current()
+	if err != nil {
+		log.Fatalf("%s", err)
+	}
+
+	pulumi.Run(func(ctx *pulumi.Context) error {
+		conf := config.New(ctx, "")
+		var clustersCfg ConfigClusters
+		conf.RequireObject("clusters", &clustersCfg)
+		DefaultNamePrefix = fmt.Sprintf(
+			"%s-%s",
+			conf.Require("resourcePrefix"),
+			u.Username,
+		)
+		cfg := Config{
+			Clusters: clustersCfg,
+			Location: conf.Require("location"),
+		}
+
+		var privateKey string
+		var publicKey string
+
+		privateKeySecret, err := secretmanager.LookupSecretVersion(ctx, &secretmanager.LookupSecretVersionArgs{
+			Secret: PrivateKeySecretName,
+		})
+		if err != nil {
+			if privateKey, publicKey, err = GenerateSSHKeys(); err != nil {
+				return err
+			}
+
+			_ = ctx.Log.Info("No ssh keys in GCP Secret Manager", nil)
+			_ = ctx.Log.Info("To create them:", nil)
+			_ = ctx.Log.Info("###", nil)
+
+			privateKey = strings.ReplaceAll(privateKey, "\n", "\\n")
+
+			createPrivKey := fmt.Sprintf("printf -- '%s' | gcloud secrets create %s --data-file=-", privateKey, PrivateKeySecretName)
+			createPubKey := fmt.Sprintf("printf -- '%s' | gcloud secrets create %s --data-file=-", publicKey, PublicKeySecretName)
+
+			_ = ctx.Log.Info(createPrivKey, nil)
+			_ = ctx.Log.Info("###", nil)
+			_ = ctx.Log.Info(createPubKey, nil)
+
+			return errors.New("cannot proceed without necessary ssh keys")
+		} else {
+			publicKeySecret, err := secretmanager.LookupSecretVersion(ctx, &secretmanager.LookupSecretVersionArgs{
```

Comment:
Nit drop the else and outdent its contents.

Reply:
Addressed